product_contract/models/sale_order.py

```diff
@@ -0,0 +1,147 @@
+# Copyright 2017 LasLabs Inc.
+# Copyright 2018 ACSONE SA/NV.
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from odoo import _, fields, api, models
+from odoo.exceptions import ValidationError
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    is_contract = fields.Boolean(string="Is a contract", compute="_compute_is_contract")
+    contract_count = fields.Integer(compute="_compute_contract_count")
+    need_contract_creation = fields.Boolean(compute="_compute_need_contract_creation")
+
+    @api.constrains("state")
+    def check_contact_is_not_terminated(self):
+        for rec in self:
+            if (
+                rec.state
+                not in (
+                    "sale",
+                    "done",
+                    "cancel",
+                )
+                and rec.order_line.filtered("contract_id.is_terminated")
+            ):
+                raise ValidationError(
+                    _("You can't upsell or downsell a terminated contract")
+                )
+
+    @api.depends("order_line.contract_id", "state")
+    def _compute_need_contract_creation(self):
+        for rec in self:
+            if rec.state in ("sale", "done"):
+                line_to_create_contract = rec.order_line.filtered(
+                    lambda r: not r.contract_id and r.product_id.is_contract
+                )
+                line_to_update_contract = rec.order_line.filtered(
+                    lambda r: r.contract_id
+                    and r.product_id.is_contract
+                    and r
+                    not in r.contract_id.contract_line_ids.mapped("sale_order_line_id")
+                )
+                if line_to_create_contract or line_to_update_contract:
+                    rec.need_contract_creation = True
```

Comment:
if above conditions are not met, there wont be value assigned to this field, probably that's why you have an error